File: go/tests/gaming_board_test.go
// gaming_board_test.go
//
// Verifies the CircleAI.Gaming port (gaming_board.go): title add/get, titles by
// genre (sorted), total play time, achievements newest-first, and most-played
// top-K ranking.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestGaming_TitlesAndPlayTime(t *testing.T) {
	b := circleai.NewInMemoryGamingBoard()
	b.AddTitle(circleai.GameTitle{TitleId: "g2", Name: "Beta", Genre: "RPG", Platform: "PC"})
	b.AddTitle(circleai.GameTitle{TitleId: "g1", Name: "Alpha", Genre: "rpg", Platform: "PC"})
	b.AddTitle(circleai.GameTitle{TitleId: "g3", Name: "Gamma", Genre: "FPS", Platform: "PC"})
	if got, ok := b.GetTitle("g1"); !ok || got.Name != "Alpha" {
		t.Fatalf("get title = %+v ok=%v", got, ok)
	}
	rpg := b.TitlesByGenre("RPG")
	if len(rpg) != 2 || rpg[0].TitleId != "g1" || rpg[1].TitleId != "g2" {
		t.Fatalf("titles-by-genre sorted failed: %+v", rpg)
	}

	now := time.Now().UTC()
	b.RecordSession(circleai.PlaySession{SessionId: "s1", UserId: "u1", TitleId: "g1", Duration: 30 * time.Minute, AtUtc: now})
	b.RecordSession(circleai.PlaySession{SessionId: "s2", UserId: "u1", TitleId: "g1", Duration: 90 * time.Minute, AtUtc: now})
	if tp := b.TotalPlayTime("u1", "g1"); tp != 120*time.Minute {
		t.Fatalf("total play time = %v, want 2h", tp)
	}
}

func TestGaming_AchievementsAndMostPlayed(t *testing.T) {
	b := circleai.NewInMemoryGamingBoard()
	b.AddTitle(circleai.GameTitle{TitleId: "g1", Name: "Alpha"})
	b.AddTitle(circleai.GameTitle{TitleId: "g2", Name: "Beta"})
	now := time.Now().UTC()
	b.Unlock(circleai.AchievementUnlock{UnlockId: "x1", UserId: "u1", TitleId: "g1", Achievement: "First", AtUtc: now.Add(-time.Hour)})
	b.Unlock(circleai.AchievementUnlock{UnlockId: "x2", UserId: "u1", TitleId: "g1", Achievement: "Second", AtUtc: now})
	ach := b.AchievementsFor("u1")
	if len(ach) != 2 || ach[0].UnlockId != "x2" {
		t.Fatalf("achievements newest-first failed: %+v", ach)
	}

	b.RecordSession(circleai.PlaySession{SessionId: "s1", UserId: "u1", TitleId: "g1", Duration: 30 * time.Minute, AtUtc: now})
	b.RecordSession(circleai.PlaySession{SessionId: "s2", UserId: "u1", TitleId: "g2", Duration: 120 * time.Minute, AtUtc: now})
	top := b.MostPlayed("u1", 5)
	if len(top) != 2 || top[0].TitleId != "g2" || top[1].TitleId != "g1" {
		t.Fatalf("most-played ranking failed: %+v", top)
	}
	if one := b.MostPlayed("u1", 1); len(one) != 1 || one[0].TitleId != "g2" {
		t.Fatalf("most-played topK failed: %+v", one)
	}
}

File: typescript/tests/gaming_board.test.ts
// gaming_board.test.ts
// Verifies the CircleAI.Gaming port: titles by genre, total play time,
// achievements newest-first, most-played ranking.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryGamingBoard,
  GamingDomainContext,
  gameTitle,
  playSession,
  achievementUnlock,
} from "../src/gaming/index";

describe("InMemoryGamingBoard", () => {
  it("adds titles and finds by genre case-insensitively", () => {
    const b = new InMemoryGamingBoard();
    b.addTitle(gameTitle("t1", "Speedster", "Racing", "PC"));
    b.addTitle(gameTitle("t2", "Puzzler", "Puzzle", "Switch"));
    assert.equal(b.getTitle("t1")?.name, "Speedster");
    assert.deepEqual(
      b.titlesByGenre("racing").map((t) => t.titleId),
      ["t1"],
    );
  });

  it("totals play time in milliseconds for a (user, title)", () => {
    const b = new InMemoryGamingBoard();
    b.recordSession(playSession("s1", "u1", "t1", 600_000, new Date("2026-01-01T00:00:00Z")));
    b.recordSession(playSession("s2", "u1", "t1", 300_000, new Date("2026-01-02T00:00:00Z")));
    b.recordSession(playSession("s3", "u1", "t2", 999_999, new Date("2026-01-02T00:00:00Z")));
    assert.equal(b.totalPlayTime("u1", "t1"), 900_000);
  });

  it("lists achievements newest-first", () => {
    const b = new InMemoryGamingBoard();
    b.unlock(achievementUnlock("x1", "u1", "t1", "First Win", new Date("2026-01-01T00:00:00Z")));
    b.unlock(achievementUnlock("x2", "u1", "t1", "Marathon", new Date("2026-01-05T00:00:00Z")));
    b.unlock(achievementUnlock("x3", "u2", "t1", "Other", new Date("2026-01-06T00:00:00Z")));
    assert.deepEqual(
      b.achievementsFor("u1").map((u) => u.unlockId),
      ["x2", "x1"],
    );
  });

  it("ranks most-played titles by total time, dropping unknown titles, capped by topK", () => {
    const b = new InMemoryGamingBoard();
    b.addTitle(gameTitle("t1", "A", "g", "PC"));
    b.addTitle(gameTitle("t2", "B", "g", "PC"));
    // t3 has sessions but no registered title → dropped.
    b.recordSession(playSession("s1", "u1", "t1", 100, new Date("2026-01-01T00:00:00Z")));
    b.recordSession(playSession("s2", "u1", "t2", 500, new Date("2026-01-01T00:00:00Z")));
    b.recordSession(playSession("s3", "u1", "t3", 9999, new Date("2026-01-01T00:00:00Z")));
    assert.deepEqual(
      b.mostPlayed("u1").map((t) => t.titleId),
      ["t2", "t1"],
    );
    assert.deepEqual(
      b.mostPlayed("u1", 1).map((t) => t.titleId),
      ["t2"],
    );
  });

  it("mostPlayed throws on non-positive topK", () => {
    const b = new InMemoryGamingBoard();
    assert.throws(() => b.mostPlayed("u1", 0), /topK/);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(GamingDomainContext.systemPromptSnippet.includes("[DOMAIN: Gaming]"));
    assert.deepEqual(GamingDomainContext.complianceFlags, ["POPIA", "WASPA", "Child_Protection"]);
    assert.deepEqual(GamingDomainContext.suggestedTools, ["game_db", "community_tools", "analytics", "web_search"]);
  });
});
